Implemented the additional properties resolver

cmodijk opened the pull request:
RabbitMQ 3.5 has a method to send a message with different priority's. But currently we can't send additional properties to the consumer. This implements my pull request https://github.com/SimpleBus/Asynchronous/pull/2.

Our use case is that we have a CommandHandler which is generating video's from images. We have a cronjob that gets new images from different api's and from that it generates images. But sometimes a customer needs de latest video fast and with the priority option it can skip the line and proces faster. 

This pull request depends on https://github.com/SimpleBus/Asynchronous/pull/2

cmodijk committed:
Implemented the additional properties resolver

cmodijk commented:
Travis-CI fails becues it depends on my other pull request and I don't know how to fix that

matthiasnoback commented:
Nice! As mentioned in your other PR, it would be great to have multiple
resolvers. You should be able to register them with a tag and a priority.
Would you agree with that?
Op 18 apr. 2015 14:25 schreef "Cliff Odijk" notifications@github.com:

> RabbitMQ 3.5 has a method to send a message with different priority's. But
> currently we cant send additional properties to the consumer. I implemented
> the base classes here but did not give an implementation like the routing
> key resolver.
> 
> Our use case is that we have a CommandHandler which is generating video's
> from images. We have a cronjob that gets new images from different api's
> and from that it generates images. But sometimes a customer needs de latest
> video fast and with the priority option it can skip the line and proces
> faster.
> 
> This pull request depends on SimpleBus/Asynchronous#2
> 
> ## https://github.com/SimpleBus/Asynchronous/pull/2
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/SimpleBus/RabbitMQBundleBridge/pull/9
> Commit Summary
> - Implemented the additional properties resolver
> 
> File Changes
> - _M_ src/DependencyInjection/Configuration.php
>   https://github.com/SimpleBus/RabbitMQBundleBridge/pull/9/files#diff-0
>   (4)
> - _M_
>   src/DependencyInjection/SimpleBusRabbitMQBundleBridgeExtension.php
>   https://github.com/SimpleBus/RabbitMQBundleBridge/pull/9/files#diff-1
>   (14)
> - _M_ src/RabbitMQPublisher.php
>   https://github.com/SimpleBus/RabbitMQBundleBridge/pull/9/files#diff-2
>   (13)
> - _M_ src/Resources/config/commands.yml
>   https://github.com/SimpleBus/RabbitMQBundleBridge/pull/9/files#diff-3
>   (1)
> - _M_ src/Resources/config/events.yml
>   https://github.com/SimpleBus/RabbitMQBundleBridge/pull/9/files#diff-4
>   (1)
> - _A_ src/Resources/config/properties.yml
>   https://github.com/SimpleBus/RabbitMQBundleBridge/pull/9/files#diff-5
>   (4)
> - _M_ tests/RabbitMQPublisherTest.php
>   https://github.com/SimpleBus/RabbitMQBundleBridge/pull/9/files#diff-6
>   (16)
> 
> Patch Links:
> - https://github.com/SimpleBus/RabbitMQBundleBridge/pull/9.patch
> - https://github.com/SimpleBus/RabbitMQBundleBridge/pull/9.diff
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/SimpleBus/RabbitMQBundleBridge/pull/9.

cmodijk commented:
Done! Love to hear any feedback you have.

matthiasnoback commented:
Looks good!

Some comments:
- Could you add a property resolver as a service for the functional test and verify that the right properties are being added?
- Maybe we should already release a new version simplebus/asynchronous so we can make the tests on Travis pass ;)

cmodijk commented:
> Could you add a property resolver as a service for the functional test and verify that the right properties are being added?

I'm not sure what you mean but i will try to add a new test inside the functional test class. 

> Maybe we should already release a new version simplebus/asynchronous so we can make the tests on Travis pass ;)

I was thinking about this i really like the idea of separating packages because with that you don't have the optional dependencies like in your blog. But doing pull requests like these you need to create multiple pull requests which have dependencies on each other how could we solve these problems without releasing things and still be able to create pull request that won't fail?

cmodijk commented:
> Could you add a property resolver as a service for the functional test and verify that the right properties are being added?

I'm trying to test this and currently i have added the service inside the functional test `config.yml` but the service is private how could i test if the resolver is registering the right properties?

matthiasnoback commented:
> I was thinking about this i really like the idea of separating packages because with that you don't have the optional dependencies like in your blog. But doing pull requests like these you need to create multiple pull requests which have dependencies on each other how could we solve these problems without releasing things and still be able to create pull request that won't fail? 

Right, this is where I regret that I didn't create one big repository and letting them be split into several smaller ones by some automatic tool. That would have made life so much easier...

In the past I did some tricks with having branch aliases in Composer, lowering minimum stability, etc. but I'm not sure it'll work in this case.

matthiasnoback commented:
Also, about the test: the unit test for the `RabbitMqPublisher` is just fine. If you can think of a way in which to functionally test this, i.e. verify something on the consuming side, etc. that would be a nice to have. Otherwise, if you see this working properly, then we could also just leave it at that.

cmodijk commented:
> In the past I did some tricks with having branch aliases in Composer, lowering minimum stability, etc. but I'm not sure it'll work in this case.

I'm doing this locally to run the test suit but i don't think i should push to my pull request right? 

> Also, about the test: the unit test for the RabbitMqPublisher is just fine. If you can think of a way in which to functionally test this, i.e. verify something on the consuming side, etc. that would be a nice to have. Otherwise, if you see this working properly, then we could also just leave it at that.

I just added a test using an alias of the delegate resolver service to make it public and then test if properties are being handled.

matthiasnoback commented:
> I just added a test using an alias of the delegate resolver service to make it public and then test if properties are being handled.

I see, you're testing that the specialized property resolver is actually used. It would also be great to verify the behavior of the publisher in some way, but this may not be easy. For now I'd say we're fine.

matthiasnoback commented:
I just tagged v1.1.0 of `simple-bus/asynchronous` by the way. So now you can change the required version in the `composer.json` to `^1.1.0` to make sure that when people install this library, those classes exist.

cmodijk commented:
> I see, you're testing that the specialized property resolver is actually used. It would also be great to verify the behavior of the publisher in some way, but this may not be easy. For now I'd say we're fine.

What do you think of this new test i just implemented. During runtime i'm changing the producer to a mocked version to capture the date being send to the producer.

> I just tagged v1.1.0 of simple-bus/asynchronous by the way. So now you can change the required version in the composer.json to ^1.1.0 to make sure that when people install this library, those classes exist.

I have updated my pull request to `"simple-bus/asynchronous": "~1.1"`

matthiasnoback commented:
Maybe you saw it but I released a new major version of this package. I removed the Message interface and all type-hints for it so maybe we can release your feature is 2.1.0?

cmodijk commented:
Yes I saw this, is there any documentation why you did this change? 

Ps; I will update my pull request.

cmodijk commented:
I have just updated my pull request to match the upstream

matthiasnoback commented:
Thanks again!

I added some documentation as well: http://simplebus.github.io/RabbitMQBundleBridge/doc/additional_properties.html and http://simplebus.github.io/Asynchronous/doc/additional_properties.html

cmodijk commented:
Cool! Tnx for writing the docs that's not my strong point.

matthiasnoback commented:
No problem!
Op 6 mei 2015 12:27 schreef "Cliff Odijk" notifications@github.com:

> Cool! Tnx for writing the docs that's not my strong point.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/SimpleBus/RabbitMQBundleBridge/pull/9#issuecomment-99410519
> .
